File: frontend/src/api/webhook.ts
/**
 * Webhook API（对齐 D05 §11）
 */
import { get, post, put, del } from './request'
import type { PageQuery, PageResult } from '@/types/api'
import type { WebhookStatus, WebhookDeliveryStatus } from '@/types/enum'

/** Webhook 列表项（D05 §11.2） */
export interface WebhookListItem {
  id: string
  url: string
  events: string[]
  status: WebhookStatus
  created_at: string
  last_delivery_at?: string
  last_delivery_status?: WebhookDeliveryStatus
}

/** Webhook 详情（D05 §11.3） */
export interface WebhookDetail extends WebhookListItem {
  secret_hash: string
  updated_at: string
  recent_deliveries: WebhookDelivery[]
}

/** 投递记录（D05 §11.7） */
export interface WebhookDelivery {
  delivery_id: string
  event_id: string
  event_type: string
  webhook_id: string
  status: WebhookDeliveryStatus
  is_test: boolean
  attempts: WebhookDeliveryAttempt[]
  delivered_at?: string
  dead_lettered_at?: string
  dead_letter_reason?: string
}

export interface WebhookDeliveryAttempt {
  attempt_no: number
  sent_at: string
  response_code: number
  response_body_snippet?: string
  latency_ms?: number
  next_retry_at?: string | null
}

/** 注册 Webhook（D05 §11.1） */
export function createWebhook(payload: {
  url: string
  events: string[]
  secret: string
  challenge_expected?: boolean
}) {
  return post<WebhookListItem>('/webhooks', payload, {
    headers: { 'X-Idempotency-Key': crypto.randomUUID() }
  })
}

/** Webhook 列表（D05 §11.2） */
export function listWebhooks(query?: PageQuery) {
  return get<PageResult<WebhookListItem>>('/webhooks', query as Record<string, unknown>)
}

/** Webhook 详情（D05 §11.3） */
export function getWebhook(id: string) {
  return get<WebhookDetail>(`/webhooks/${id}`)
}

/** 更新 Webhook（D05 §11.4） */
export function updateWebhook(
  id: string,
  payload: { url: string; events: string[]; secret?: string; challenge_expected?: boolean }
) {
  return put<WebhookListItem>(`/webhooks/${id}`, payload)
}

/** 注销 Webhook（D05 §11.5） */
export function deleteWebhook(id: string) {
  return del<void>(`/webhooks/${id}`)
}

/** 测试投递（D05 §11.6） */
export function testWebhook(id: string, payload: { event_type: string; test_payload?: Record<string, unknown> }) {
  return post<{ delivery_id: string; status: string; signature_header: string }>(
    `/webhooks/${id}/test`,
    payload
  )
}

/** 投递记录（D05 §11.7） */
export function listWebhookDeliveries(
  id: string,
  query: PageQuery & {
    event_type?: string
    status?: WebhookDeliveryStatus
    start_time?: string
    end_time?: string
  }
) {
  return get<PageResult<WebhookDelivery>>(`/webhooks/${id}/deliveries`, query as Record<string, unknown>)
}

File: backend/app/schemas/webhook.py
"""Webhook schemas（D05 §11）。"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field, field_validator

from app.services.webhook import validate_webhook_secret, validate_webhook_url


class WebhookCreate(BaseModel):
    """POST /webhooks 请求体（D05 §11.1）。

    merchant_id 显式指定配置写入的目标商户行；缺失/非法由端点
    返回 400（INVALID_PARAMS），不走 Pydantic 必填（避免 422 语义）。
    """

    merchant_id: str | None = None
    url: str
    events: list[str]
    secret: str
    challenge_expected: bool = True

    @field_validator("url")
    @classmethod
    def _check_url(cls, v: str) -> str:
        return validate_webhook_url(v)

    @field_validator("secret")
    @classmethod
    def _check_secret(cls, v: str) -> str:
        return validate_webhook_secret(v)


class WebhookUpdate(BaseModel):
    """PUT /webhooks/{id} 请求体（D05 §11.4）。

    merchant_id 必须与路径 {id} 一致（缺失/不一致端点返回 400）。
    secret 可选：省略时保留原签名密钥（仅改 URL/事件无需重输 secret）。
    """

    merchant_id: str | None = None
    url: str
    events: list[str]
    secret: str | None = None
    challenge_expected: bool = True

    @field_validator("url")
    @classmethod
    def _check_url(cls, v: str) -> str:
        return validate_webhook_url(v)

    @field_validator("secret")
    @classmethod
    def _check_secret(cls, v: str | None) -> str | None:
        if v is None:
            return None
        return validate_webhook_secret(v)


class WebhookOut(BaseModel):
    """Webhook 详情。"""

    id: str
    url: str
    events: list[str] = Field(default_factory=list)
    status: str = "PENDING_VERIFICATION"
    secret_hash: str | None = None
    created_at: datetime
    updated_at: datetime | None = None
    last_delivery_at: datetime | None = None
    last_delivery_status: str | None = None
    challenge_id: str | None = None


class WebhookTestRequest(BaseModel):
    """POST /webhooks/{id}/test 请求体（D05 §11.6）。"""

    event_type: str
    test_payload: dict[str, Any] | None = None


class WebhookTestResponse(BaseModel):
    """POST /webhooks/{id}/test 响应。"""

    delivery_id: str
    webhook_id: str
    event_type: str
    status: str = "PENDING"
    signature_header: str


class DeliveryAttempt(BaseModel):
    """投递尝试记录。"""

    attempt_no: int
    sent_at: datetime
    response_code: int | None = None
    response_body_snippet: str | None = None
    latency_ms: int | None = None
    next_retry_at: datetime | None = None


class WebhookDeliveryOut(BaseModel):
    """投递记录（D05 §11.7）。"""

    delivery_id: str
    event_id: str
    event_type: str
    webhook_id: str
    status: str
    is_test: bool = False
    attempts: list[DeliveryAttempt] = Field(default_factory=list)
    delivered_at: datetime | None = None
    dead_lettered_at: datetime | None = None
    dead_letter_reason: str | None = None


__all__ = [
    "DeliveryAttempt",
    "WebhookCreate",
    "WebhookDeliveryOut",
    "WebhookOut",
    "WebhookTestRequest",
    "WebhookTestResponse",
    "WebhookUpdate",
]
